### PR TITLE
Fix streaming

### PR DESCRIPTION
**Pull request by Disiok**
### Summary
* Fix streaming support for langchain LLMs

### Details
* previously we rely on `model.stream()`. This is now deprecated (and broken) in langchain, also lack chat model support.
* now we adapt to langchain new callback mechanism by:
  1. running the model predict in separate thread (to avoid blocking)
  2. use a callback handler to pull tokens into a queue
  3. return a generator to allow downstream users to iterate over tokens as they arrive

### Dirty laundry
* we override `__deepcopy__` for our callback handler to avoid being deepcopied in langchain callback manager code
* we silently attach streaming callback handler to the LLM

### CODE REVIEW

File: llama_index/llm_predictor/base.py [hunk 1]
```diff
@@ -250,11 +255,21 @@ def stream(self, prompt: Prompt, **prompt_args: Any) -> Tuple[Generator, str]:
...
         formatted_prompt = prompt.format(llm=self._llm, **prompt_args)
-        raw_response_gen = self._llm.stream(formatted_prompt)
-        response_gen = _get_response_gen(raw_response_gen)
+
+        handler = StreamingGeneratorCallbackHandler()
+
+        assert isinstance(self._llm, (BaseLLM, BaseChatModel))
+        self._llm.callbacks = [handler]
+
+        assert isinstance(self._llm, (BaseOpenAI, ChatOpenAI))
```
Comment (logan-markewich) [inline llama_index/llm_predictor/base.py:265]: So this means streaming will only work for openai models? In the langchain code, the callback handler should work for any LLM that uses their run manager + on_llm_new_token 🤔

**Review by logan-markewich — APPROVED** [review]
Works well enough for me! Minor, but we could also add a page for streaming somewhere in the docs. Otherwise the notebooks are good enough

File: llama_index/llm_predictor/base.py [hunk 2]
```diff
@@ -250,11 +255,21 @@ def stream(self, prompt: Prompt, **prompt_args: Any) -> Tuple[Generator, str]:
...
+        self._llm.streaming = True
```
Comment (jerryjliu) [inline llama_index/llm_predictor/base.py:266]: could also just error if _llm.streaming was set to False

File: llama_index/llm_predictor/base.py (hunk 1 above)
Comment (jerryjliu) [reply llama_index/llm_predictor/base.py:265]: yeah agreed here, be explicit that streaming only supported for openai llm's atm

**Review by jerryjliu — APPROVED** [review]

**Comment by jerryjliu** [comment]
> Works well enough for me! Minor, but we could also add a page for streaming somewhere in the docs. Otherwise the notebooks are good enough

definitely agreed with adding a page here

File: llama_index/llm_predictor/base.py (hunk 2 above)
Comment (Disiok) [reply llama_index/llm_predictor/base.py]: Revised

File: llama_index/llm_predictor/base.py (hunk 1 above)
Comment (Disiok) [reply llama_index/llm_predictor/base.py]: revised

**Pull request merged**